src/compiler/commandLineParser.ts

```diff
@@ -1212,94 +1511,223 @@ namespace ts {
      */
     function parseConfig(
             json: any,
+            sourceFile: JsonSourceFile,
             host: ParseConfigHost,
             basePath: string,
             configFileName: string,
             resolutionStack: Path[],
             errors: Diagnostic[],
-            ): ParsedTsconfig {
-
+    ): ParsedTsconfig {
         basePath = normalizeSlashes(basePath);
         const getCanonicalFileName = createGetCanonicalFileName(host.useCaseSensitiveFileNames);
         const resolvedPath = toPath(configFileName || "", basePath, getCanonicalFileName);
 
         if (resolutionStack.indexOf(resolvedPath) >= 0) {
             errors.push(createCompilerDiagnostic(Diagnostics.Circularity_detected_while_resolving_configuration_Colon_0, [...resolutionStack, resolvedPath].join(" -> ")));
-            return { include: undefined, exclude: undefined, files: undefined, options: {}, compileOnSave: undefined };
+            return { raw: json || convertToObject(sourceFile, errors) };
+        }
+
+        const ownConfig = json ?
+            parseOwnConfigOfJson(json, host, basePath, getCanonicalFileName, configFileName, errors) :
+            parseOwnConfigOfJsonSourceFile(sourceFile, host, basePath, getCanonicalFileName, configFileName, errors);
+
+        if (ownConfig.extendedConfigPath) {
+            // copy the resolution stack so it is never reused between branches in potential diamond-problem scenarios.
+            resolutionStack = resolutionStack.concat([resolvedPath]);
+            const extendedConfig = getExtendedConfig(sourceFile, ownConfig.extendedConfigPath, host, basePath, getCanonicalFileName,
+                resolutionStack, errors);
+            if (extendedConfig && isSuccessfulParsedTsconfig(extendedConfig)) {
+                const baseRaw = extendedConfig.raw;
+                const raw = ownConfig.raw;
+                const setPropertyInRawIfNotUndefined = (propertyName: string) => {
+                    const value = raw[propertyName] || baseRaw[propertyName];
+                    if (value) {
+                        raw[propertyName] = value;
+                    }
+                };
+                setPropertyInRawIfNotUndefined("include");
+                setPropertyInRawIfNotUndefined("exclude");
+                setPropertyInRawIfNotUndefined("files");
+                if (raw.compileOnSave === undefined) {
+                    raw.compileOnSave = baseRaw.compileOnSave;
+                }
+                ownConfig.options = assign({}, extendedConfig.options, ownConfig.options);
+                // TODO extend type typeAcquisition

```

Comment:
I find it useful to use no space on TODO comments so the linter catches them. E.g. `//extend typeAcquisition`.